lib/active_shipping/package.rb

```diff
@@ -1,118 +1,62 @@
+    def initialize(weight:, length:, width:, height: nil, options: {})
+      @options = @@default_options.merge(options.symbolize_keys)
 
-      imperial = (options[:units] == :imperial) ||
-                 ([grams_or_ounces, *dimensions].all? { |m| m.respond_to?(:unit) && m.unit.to_sym == :imperial })
+      raise ArgumentError, "Weight needs to be a Measured::Measurable object" unless weight.is_a?(Measured::Measurable)
+      raise ArgumentError, "Length needs to be a Measured::Measurable object" unless length.is_a?(Measured::Measurable)
+      raise ArgumentError, "Width needs to be a Measured::Measurable object" unless width.is_a?(Measured::Measurable)
+      raise ArgumentError, "Height needs to be a Measured::Measurable object" unless height.nil? || length.is_a?(Measured::Measurable)
 
-      weight_imperial = dimensions_imperial = imperial if options.include?(:units)
+      @weight = weight
 
-      if options.include?(:weight_units)
-        weight_imperial = (options[:weight_units] == :imperial) ||
-                          (grams_or_ounces.respond_to?(:unit) && m.unit.to_sym == :imperial)
-      end
-
-      if options.include?(:dim_units)
-        dimensions_imperial = (options[:dim_units] == :imperial) ||
-                              (dimensions && dimensions.all? { |m| m.respond_to?(:unit) && m.unit.to_sym == :imperial })
-      end
-
-      @weight_unit_system = weight_imperial ? :imperial : :metric
-      @dimensions_unit_system = dimensions_imperial ? :imperial : :metric
-
-      @weight = attribute_from_metric_or_imperial(grams_or_ounces, Measured::Weight, @weight_unit_system, :grams, :ounces)
+      @length = length
+      @width  = width
+      @height = height || ( cylinder? ? width : Measured::Length.new(0, length.unit) )
 
-      if @dimensions.blank?
-        zero_length = Measured::Length.new(0, (dimensions_imperial ? :inches : :centimetres))
-        @dimensions = [zero_length] * 3
-      else
-        process_dimensions
-      end
+      @value = self.class.cents_from(options[:value])
+      @currency = options[:currency] || options[:value].try(:currency)
+    end
 
-      @value = Package.cents_from(options[:value])
-      @currency = options[:currency] || (options[:value].currency if options[:value].respond_to?(:currency))
-      @cylinder = (options[:cylinder] || options[:tube]) ? true : false
-      @gift = options[:gift] ? true : false
-      @oversized = options[:oversized] ? true : false
-      @unpackaged = options[:unpackaged] ? true : false
+    def dimensions
+      [@length, @width, @height]
     end
 
-    def unpackaged?
-      @unpackaged
+    def girth
+      @girth ||= cylinder ? width.scale(Math::PI) : (width + height).scale(2)
     end
+    alias_method :circumference, :girth
+    alias_method :around, :girth
 
-    def oversized?
-      @oversized
+    def volume
+      @volume ||= if cylinder?
+        Math::PI * width.scale(0.5).value * height.scale(0.5).value * length.value
```

Comment:
Two things:

First, would the measured gem value from a function to coerce a list of values into the same unit? For example, this would be written as:

```ruby
Measured.coerce_unit(width.scale(0.5), height.scale(0.5), length).reduce(&:*)
```

We could also add a function to do the inner product of the values until we support multiplying these together.

Second, I would recommend parentheses around the three terms on the right, because the first multiply will coerce things into a float and then you'll lose some precision gained from `:value` being `Rational` or `BigDecimal`. The other option would be to create a `Rational` or `BigDecimal` PI constant in this class and use that instead (recommended). The nice thing about that is that it will reduce the number of return value types.